source/base/mpi.cc

```diff
@@ -113,6 +113,33 @@ namespace Utilities
 
 
 
+    CriticalSection::CriticalSection(const MPI_Comm &comm, MPI_Request &request)
+      : comm(comm)
+      , request(request)
+    {
+      Utilities::MPI::MPI_InitFinalize::register_static_request(request);
+      wait();
+    }
+
+
+
+    void
+    CriticalSection::wait()
+    {
+      const int ierr = MPI_Wait(&request, MPI_STATUS_IGNORE);
+      AssertThrowMPI(ierr);
+    }
+
+
+
+    CriticalSection::~CriticalSection()
+    {
+      const int ierr = MPI_Ibarrier(comm, &request);
+      AssertThrowMPI(ierr);
```

Comment:
```suggestion
#ifdef DEAL_II_WITH_MPI
#  if DEAL_II_MPI_VERSION_GTE(3, 0)
      const int ierr = MPI_Ibarrier(comm, &request);
      AssertThrowMPI(ierr);
# else 
      const int ierr = MPI_Barrier(comm);
      AssertThrowMPI(ierr);
#  endif
#endif
```

Reply:
Out of curiosity, what's the fraction of systems out there that don't yet have MPI 3?